pandas/_libs/lib.pyx

```diff
@@ -1203,9 +1204,14 @@ cdef object _try_infer_map(object dtype):
         object val
         str attr
     for attr in ["name", "kind", "base"]:
-        val = getattr(dtype, attr)
+        val = getattr(dtype, attr, None)
         if val in _TYPE_MAP:
             return _TYPE_MAP[val]
+        # also check base name for parametrized dtypes (eg period[D])
+        if isinstance(val, str):
+            val = val.split("[")[0]
```

Comment:
can we explicity add the Period[D] types to the type_map (there aren't that many)

Reply:
Going from the ``PeriodDtypeCode`` enum, there are actually quite some?

Reply:
if we were to get that specific, at some point it makes more sense to just return a dtype object

Reply:
right here why can't we try to convert to an EA type using `registry.find()`?

Reply:
Because we already have a `dtype` here, there is no need to look anything up in the registry. We have an EA `dtype` that we want to find in the TYPE_MAP, by checking the name, kind or base attributes of the dtype.

Reply:
no you don't, its a string! pls convert to a dtype (before you check if its in the type map), then we can move forward

Reply:
The function we are commenting on is `_try_infer_map(object dtype)`, i.e. it receives a dtype. And see a few lines above this comment, line 1230, we get the `val` as an attribute from the `dtype`. I can ensure you we start here from a dtype

Reply:
`val` is of course a string, to be clear, but we got `val` as the attribute *of a dtype*, so there is no need to convert it back to a dtype

Reply:
I am so -1 on parsing the dtype string like this. this is a gigantic hack. pls find some other way to do this.

Reply:
If you have a specific proposal, I am all ears. But this is the simplest solution I could come up with for now. Parameterized dtypes is something quite common, and they all use the `name[param]` pattern, so this is just a line of code to get the "name" part of the dtype's full name. 
I think that's a quite general solution (an alternative could be to expose this "base" name as a new attribute on the dtype)

Reply:
so i have offered multiple solns. convert it to a dtype object as its a string! then you *could* if you need have a method to instrospect what is needed for TYPE_MAP (but i think that is just orthogonal here), you can prob just get the base dtype.

Reply:
my point is this is re-inventing (in a very hackish way) the dtype string parsing.

Reply:
> so i have offered multiple solns. convert it to a dtype object as its a string!

Again, we *already have a dtype object* .. The whole point is to go from a dtype object to the string representation returned by `infer_dtypes`. 

> then you could if you need have a method to instrospect what is needed for TYPE_MAP

That is exactly what I am doing here

Reply:
@jorisvandenbossche is right here.  e.g. for `datetime64[ns, tz]` we _would_ have to do this string parsing if its `dtype.kind` wasn't hard-coded in _TYPE_MAP.

> you can prob just get the base dtype.

The trouble is that for PeriodDtype or IntervalDtype the base dtype is object (or non-existent)

Reply:
@jreback any further comment here?

Reply:
i haven't changed my mind
parsing here the way you are doing is -1

Reply:
welcome to have a comprehensive soln - but this is way too hacky and fragile

Reply:
To repeat myself from above: if you have a specific proposal how this can be done better, I am all ears. But this is the simplest solution I could come up with for now. Brock seems to be fine with it as well.

(and please don't say that I should convert the string to a dtype, as you can see in the code a few lines above, we actually start from a dtype object)

Reply:
I guess a stopgap solution could be for `_try_infer_map` to also check `dtype.type` and to add `{Interval: "interval", Decimal: "decimal", Period: "period"}` to `_TYPE_MAP`?

Reply:
Do we want to import those in lib.pyx ? (might be fine, eg there are already imports from tslibs.period, but at the moment only `cimport` s) 

If that's OK, I am fine with changing it. I don't necessarily find it less "hacky" than the current solution, but I just want *some* solution that is acceptable for all

Reply:
do we not already import all of the scalar EA types?
why is this any different

+1 on using the existing machinery

Reply:
>  I don't necessarily find it less "hacky" than the current solution, but I just want some solution that is acceptable for all

100% agree on both points

> Do we want to import those in lib.pyx

not ideal, but i dont think it will harm anything ATM

Reply:
> and please don't say that I should convert the string to a dtype, as you can see in the code a few lines above, we actually start from a dtype object)

exactly you are missing the entire point of the dtype abstraction

you avoid parsing strings in the first place

i will be blocking this util / unless a good soln is done

Reply:
> do we not already import all of the scalar EA types?

lib.pyx doesn't know anything about EAs. It only imports helper functions like `is_period_object` 

>  why is this any different

different than what?